doc/mpc_node_specification.md

```diff
@@ -0,0 +1,225 @@
+Each participant has its own instance of this state machine.
+
+The arrows in blue show how the `Owner` transitions between states, while the
+red arrows are for all other `Participants`. Red arrows generally only wait on
+messages by the owner, while the owner (blue) sends these messages whenever
+itself transitions forward.
+
+For the signature protocol, all participants must check that requests were
+indexed and finalized, as noted in the graph.
+
+![graph](./graphs/protocol_states.svg)
+
+
+### States
+
+`init` and `done` are implicit states. All other states must be explicitly
+stored and ideally persisted.
+
+#### Prepare
+
+In the `Prepare` state, nodes establish the protocol inputs, such as the
+participants list. Nodes should not send messages to other nodes about this
+protocol yet, nor should they process messages that they receive.
+
+Why can we not send messages, yet? This is due to the black box assumption on
+the underlying cryptography. Signing messages and sending them out might reveal
+information. This could lead to security vulnerabilities if we later invoke the
+same protocol again but with a different set of participants, for example.
+
+Nodes in the `Prepare` state may receive messages from other nodes that are
+already in the `Running` state. In that case, nodes should either reject
+messages or keep them in a buffer. If rejected, the sender (in `Running` state)
+must retry sending the rejected message. If buffered, the receiver should
+process the buffered messages right after it transitions to the `Running` state.
+
+#### Running
+
+In the `Running` state, nodes continuously poke the local message generators and
+exchange messages to progress the protocol.
+
+Messages sent to participants that are offline or reject the message must be
+retried until either the receiver accepts the message or the protocol invocation
+globally times out.
+
+
+### Owner Transitions
+
+The owner of the protocol invocation progresses the protocol:
+
+- When a new valid request is registered (indexed for the signature protocol,
+  initiated by background task for triple/pre-signature generation)
+    - Select a list of participants, based on the own view of which nodes are
+      active and hold the relevant data in their storage. We heavily rely on
+      state sync for this selection of participants to be good on the first try.
+    - Asynchronously send an INIT message to all the participants and wait until
+      all messages have been received. (HTTP success response)
+    - Transition to `Prepare` state
+- Once all INIT messages have been answered with a success response
+    - Wait for finality of the request (for signatures only)
+    - Send out a START message all participants
+    - Transition to `Running` state
+    - Start poking own generator and sending out the corresponding messages
+
+Owners must also handle the following error case:
+- Not all participants answered the INIT message with an HTTP success response
```

Comment:
Here I suggest to send INIT request to all (or at least T+1 nodes) and wait for T Nodes to respond instead of failing or retrying with a new participant set, because the timeout for INIT will slow down the protocol if one of the nodes is slow or not responding at all.

Reply:
Sounds good, yeah. Then we have to send the participant list in the START message rather than the INIT message.